### ISSUE TITLE
[Bug] trtllm-gen spec-as-decode (q_len_per_req=8) with BF16 query + FP8 KV cache silently mis-computes sliding-window attention on SM100

### ISSUE THREAD

**Issue by elad-inferize** [flashinfer-ai/flashinfer#3864]
## Summary

`trtllm-gen` spec-as-decode (`trtllm_batch_decode_with_kv_cache` with `q_len_per_req=8`) with **BF16 query + FP8 KV cache** silently mis-computes **sliding-window** attention on SM100 (B200) once `seq_len > window`. The corruption is severe (max abs err ~0.8–1.8 on outputs of magnitude ~1–3, cosine down to 0.93 vs reference) but *localized*: typically a single (query-row, head) pair per failing call, and only at some `seq_len` values (data-dependent).

Controls on the exact same inputs are all clean:
- **FP8 query** at the same (seed, seq_len): max err ~0.02 (quant noise) — clean
- **window off** (`window_left=-1`): max err ~0.004 — clean
- `q_len_per_req` 1/2/4: clean in the tested grid
- seq_len ≤ window: clean

Because the error is silent and localized, in end-to-end serving it manifests as sporadic wrong tokens / short repetition loops rather than a crash — easy to mistake for model behavior.

## Environment

- flashinfer-python **0.6.14** (latest release; flashinfer-cubin 0.6.13), torch 2.11.0+cu130, CUDA 13
- NVIDIA B200 (SM100), driver 580.x
- Geometry from gemma-4-31B: `head_dim=256`, 32 q heads, 16 kv heads, `sm_scale=1.0`, `window_left=1023`, `page_size=16`, HND

## Repro

```python
import torch
from flashinfer.decode import trtllm_batch_decode_with_kv_cache

DEV = "cuda"
FP8 = torch.float8_e4m3fn

def torch_ref(q, k, v, sm_scale, window_left, qlen):
    total = k.shape[0]
    hq, hk = q.shape[1], k.shape[1]
    rep = hq // hk
    kr = k.repeat_interleave(rep, dim=1).float().permute(1, 0, 2)
    vr = v.repeat_interleave(rep, dim=1).float().permute(1, 0, 2)
    qf = q.float().permute(1, 0, 2)
    logits = torch.bmm(qf, kr.transpose(1, 2)) * sm_scale
    qpos = torch.arange(total - qlen, total, device=q.device)
    kpos = torch.arange(total, device=q.device)
    mask = kpos[None, :] <= qpos[:, None]
    if window_left >= 0:
        mask = mask & (kpos[None, :] >= (qpos[:, None] - window_left))
    logits = logits.masked_fill(~mask[None], float("-inf"))
    return torch.bmm(torch.softmax(logits, dim=-1), vr).permute(1, 0, 2)

def run_case(total_len, qlen, q_fp8, window_left, page=16, hq=32, hk=16, hd=256,
             sm=1.0, seed=0):
    torch.manual_seed(seed)
    n_pages = (total_len + page - 1) // page
    q = torch.randn(qlen, hq, hd, device=DEV, dtype=torch.bfloat16) * 0.5
    k = torch.randn(total_len, hk, hd, device=DEV, dtype=torch.bfloat16) * 0.5
    v = torch.randn(total_len, hk, hd, device=DEV, dtype=torch.bfloat16) * 0.5
    k_store, v_store = k.to(FP8), v.to(FP8)
    k_eff, v_eff = k_store.to(torch.bfloat16), v_store.to(torch.bfloat16)
    cache = torch.zeros(n_pages, 2, hk, page, hd, device=DEV, dtype=FP8)
    for p in range(n_pages):
        lo, hi = p * page, min((p + 1) * page, total_len)
        cache[p, 0, :, : hi - lo] = k_store[lo:hi].permute(1, 0, 2)
        cache[p, 1, :, : hi - lo] = v_store[lo:hi].permute(1, 0, 2)
    q_in = q.to(FP8) if q_fp8 else q
    ws = torch.zeros(256 * 1024 * 1024, dtype=torch.uint8, device=DEV)
    out = trtllm_batch_decode_with_kv_cache(
        query=q_in, kv_cache=cache, workspace_buffer=ws,
        block_tables=torch.arange(n_pages, dtype=torch.int32, device=DEV).unsqueeze(0),
        seq_lens=torch.tensor([total_len], dtype=torch.int32, device=DEV),
        max_seq_len=total_len, bmm1_scale=sm, bmm2_scale=1.0,
        window_left=window_left,
        out=torch.empty(qlen, hq, hd, device=DEV, dtype=torch.bfloat16),
        kv_layout="HND", backend="trtllm-gen", q_len_per_req=qlen,
    )
    ref = torch_ref(q_in.to(torch.bfloat16), k_eff, v_eff, sm, window_left, qlen)
    err = (out.float() - ref).abs()
    cos = torch.nn.functional.cosine_similarity(
        out.float().flatten(1), ref.flatten(1), dim=1)
    return err, cos

for q_fp8 in (False, True):
    for wl in (1023, -1):
        for seed, total in ((0, 1286), (0, 2053), (2, 2053)):
            err, cos = run_case(total, 8, q_fp8, wl, seed=seed)
            print(f"q={'fp8' if q_fp8 else 'bf16'} wl={wl:5d} seed={seed} "
                  f"seq={total}: max_err={err.max():.4f} min_cos={cos.min():.3f}")
```

Output on B200 / flashinfer 0.6.14:

```
q=bf16 wl= 1023 seed=0 seq=1286: max_err=1.3916 min_cos=0.935   <-- BROKEN
q=bf16 wl= 1023 seed=0 seq=2053: max_err=1.7756 min_cos=0.937   <-- BROKEN
q=bf16 wl= 1023 seed=2 seq=2053: max_err=0.9783 min_cos=0.977   <-- BROKEN
q=bf16 wl=   -1 seed=0 seq=1286: max_err=0.0039 min_cos=1.000
q=bf16 wl=   -1 seed=0 seq=2053: max_err=0.0039 min_cos=1.000
q=bf16 wl=   -1 seed=2 seq=2053: max_err=0.0039 min_cos=1.000
q=fp8  wl= 1023 seed=0 seq=1286: max_err=0.0255 min_cos=1.000
q=fp8  wl= 1023 seed=0 seq=2053: max_err=0.0282 min_cos=1.000
q=fp8  wl= 1023 seed=2 seq=2053: max_err=0.0165 min_cos=1.000
q=fp8  wl=   -1 ...                                clean
```

Error structure at (seed=0, seq=1286): per-query-row max err `[0.004, 1.392, 0.011, 0.086, 0.007, 0.004, 0.004, 0.003]` — row 1 only, and within that row only head 24 exceeds 0.25. This looks like one KV tile near the window boundary being mis-handled (skipped/mis-masked) for a subset of CTAs in the BF16-Q × FP8-KV sliding-window spec-decode kernel variant.

More observations from a wider sweep:
- Sporadic in `seq_len`: e.g. at `wl=1023, qlen=8` seeds 0–2 over seq 1024–2080, failures at 1286 and 2053; at `wl=127` failures at seq 136, 137, 141, 142 (err up to 1.4).
- `page_size=32`: no failures in the same scan; `page_size=64`: fails at seq 1031 (err 1.33, cos 0.953) — page-size-dependent.
- Repeat-run deterministic (same inputs → bitwise same wrong output), and content-invariant across rows — it is a fixed indexing/predicate defect, not a race.

## End-to-end impact

Found via vLLM: serving gemma-4-31B (50/60 layers SWA window 1024) with MTP speculative decoding (verify = spec-as-decode `q_len_per_req=8`), `--kv-cache-dtype fp8_e4m3` and `--attention-config '{"disable_flashinfer_q_quantization": true}'` (BF16-Q) produces sporadic corrupted tokens and short repetition loops in otherwise coherent output, e.g. "Instead of instead of instead of instead (of using a blocking two-phase commit)". The same server without the flag (FP8-Q), or with the flag but without speculation (q_len_per_req=1), is clean. vLLM issue: (link to follow in a comment)

**Comment by elad-inferize** [comment]
vLLM-side issue with the end-to-end manifestation and the (independent) BF16-Q scale-bookkeeping bug: https://github.com/vllm-project/vllm/issues/47847

**Comment by hello-args** [comment]
I reproduced the same behavior from the vLLM side, and it lines up with the characterization here.

* Temporary downstream guard: vllm-project/vllm#47899

On SM100 with the FlashInfer backend, speculative decode routes through `trtllm_batch_decode_with_kv_cache` (`q_len_per_req > 1`). With `disable_flashinfer_q_quantization=True`, queries remain BF16 while the KV cache is FP8/NVFP4, so this exercises the same BF16-Q + quantized-KV path described in this issue.

The controls also match what I observed during serving:

* FP8-Q → clean
* No speculative decode (`q_len_per_req=1`) → clean
* Full-attention-only models → clean

The serving symptoms are also consistent: once generation crosses the sliding-window boundary, output can degrade into sporadic incorrect tokens and short repetition loops rather than crashing.

As a downstream mitigation, I opened vllm-project/vllm#47899, which adds a temporary init-time guard to reject the known unsupported configuration instead of allowing silent incorrect outputs. The guard can be removed once this issue is fixed.

Happy to validate a candidate FlashInfer fix against the vLLM serving repro if that would be helpful.

**Comment by XFDG** [comment]
Hi, I reran the issue's BF16-Q + FP8-KV sliding-window reproducer on an NVIDIA B200 (SM100) with current `main` at `c05407ceffb7d9ce111a73553a8e37ad752adb62`, CUDA 13.1, and the pinned TRTLLM-Gen FMHA artifact `2d6a5a029eefcc388ec0ceb87efb55d8bcce5c3c`.

I could not reproduce the corruption:

```
q=bf16 wl=1023 seq=1286: max_err=0.006523 min_cos=0.999997
q=bf16 wl=  -1 seq=1286: max_err=0.007327 min_cos=0.999997
q=bf16 wl=1023 seq=2053: max_err=0.007154 min_cos=0.999997
q=bf16 wl=  -1 seq=2053: max_err=0.006541 min_cos=0.999997
```

The sliding-window cases are now in the same error range as the no-window controls, unlike the reported `1.3916` / `1.7756` failures. This appears to have been fixed by subsequent kernel/artifact updates, so I do not have a justified code change to submit. If this still reproduces with current `main`, please share the exact artifact pin and I can retest that binary.
